src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs

```diff
@@ -1415,6 +1416,163 @@ public static bool VerifySign(SafeProvHandle hProv, SafeKeyHandle hKey, int calg
+            int cbPassword,
+            int dwFlags,
+            byte[] IV_Out,
+            int cbIV_In,
+            ref byte[] pbKey)
+        {
+            VerifyValidHandle(hProv);
+
+            SafeHashHandle hHash = null;
+            SafeKeyHandle hKey = null;
+            try
+            {
+                if (!Interop.CryptCreateHash(hProv, algidHash, SafeKeyHandle.InvalidHandle, CryptCreateHashFlags.None, out hHash))
+                {
+                    int hr = Marshal.GetHRForLastWin32Error();
+                    throw hr.ToCryptographicException();
+                }
+
+                // Hash the password string
+                if (!Interop.CryptHashData(hHash, password, cbPassword, 0))
+                {
+                    int hr = Marshal.GetHRForLastWin32Error();
+                    throw hr.ToCryptographicException();
+                }
+
+                // Create a block cipher session key based on the hash of the password
+                if (!Interop.CryptDeriveKey(hProv, algid, hHash, dwFlags | (int)CryptGenKeyFlags.CRYPT_EXPORTABLE, out hKey))
+                {
+                    int hr = Marshal.GetHRForLastWin32Error();
+                    throw hr.ToCryptographicException();
+                }
+
+                // Get the key contents
+                byte[] rgbKey = null;
+                int cbKey = 0;
+                UnloadKey(hProv, hKey, ref rgbKey, ref cbKey);
+
+                // Get the length of the IV
+                int cbIV = 0;
+                if (!Interop.CryptGetKeyParam(hKey, (int)CryptGetKeyParamFlags.KP_IV, null, ref cbIV, 0))
+                {
+                    int hr = Marshal.GetHRForLastWin32Error();
+                    throw hr.ToCryptographicException();
+                }
+
+                // Now allocate space for the IV
+                byte[] pbIV = new byte[cbIV];
+                if (!Interop.CryptGetKeyParam(hKey, (int)CryptGetKeyParamFlags.KP_IV, pbIV, ref cbIV, 0))
+                {
+                    int hr = Marshal.GetHRForLastWin32Error();
+                    throw hr.ToCryptographicException();
+                }
+
+                if (cbIV != cbIV_In)
+                {
+                    throw new CryptographicException(SR.Cryptography_PasswordDerivedBytes_InvalidIV);
+                }
+
+                // Copy the IV
+                Buffer.BlockCopy(pbIV, 0, IV_Out, 0, cbIV);
+
+                pbKey = new byte[cbKey];
+                Buffer.BlockCopy(rgbKey, 0, pbKey, 0, cbKey);
+            }
+            finally
+            {
+                hKey?.Dispose();
+                hHash?.Dispose();
+            }
+        }
+
+        // Helper method used by DeriveKey (above) to return the key contents.
+        // WARNING: This function side-effects its first argument (hProv)
+        private static void UnloadKey(SafeProvHandle hProv, SafeKeyHandle hKey, ref byte[] key_out, ref int cb_out)
+        {
+            byte[] rgbPubKey =
```

Comment:
Changed it to static readonly. The CryptImportKey documentation does not say it may modify.